native/opencv-wrapper.cc

```diff
@@ -116,6 +117,22 @@ void cv_vec_of_rect_drop(VecRect* v) {
     }
 }
 
+void cv_vec_of_point_drop(VecPoint* pv) {
+    if (pv->array != nullptr) {
+        free(pv->array);
+        pv->array = nullptr;
+        pv->size = 0;
+    }
+}
+
+void cv_vec_of_points_drop(VecPoints* pvs) {
+    if (pvs->array != nullptr) {
+        free(pvs->array);
```

Comment:
Here, each array element may contain a pointer (as in `VecPoint`). Wouldn't `free` here cause memory leak?

Reply:
No, `VecPoint` is `Drop` itself. We only must consider our own memory. If you try to free some `VecPoints` here you get SEGFAULT because of double memory free.

Reply:
But in Rust, it's not `Vec<VecPoint>`. It's `array: *mut CVecOfPoint`. So the `Drop` for `VecPoint` is not called because Rust doesn't know how to destroy a pointer.

Reply:
If you allocate `VecPoints` with say 10 values, and you drop it, you can see 10 calls of `VecPoint.drop` and one `VerPoints.drop`. Rust calls `drop` for all `VecPoint` itself and it destroys `VecPoints` only after it. You must not interfer with other types that implement `drop`.

Reply:
If you are in doubt, just create a test method and try to allocate lots of `VecPoints` and check if there is any leak.

Reply:
Yes, I've tested: [code here](https://github.com/nebgnahz/cv-rs/commit/876215603dbb25172352f20f849f0888e89d273b).

Running for a few seconds, the tests program has used up to 3-5G memory.

Reply:
@nebgnahz wow. Ok, need some investigation here. Sounds weird: it should free all the memory,,,

Reply:
Hmm, it seems that we must dealloc C++ code. Rust `drop` is fine.

Reply:
`CVecPoint::drop` is never called (see https://github.com/nebgnahz/cv-rs/commit/b8084b812500d9d98e6ccb43582a61e0212819c3 when I just use `unimplemented{}` there.

Reply:
@nebgnahz problem is in `vec_rect_cxx_to_c` and its counterparts because they are copying data instead of moving it. You can see the same memory leak for `cv_hog_detect`

Reply:
No, the problem is `Vec<Vec<Point>>` has two layers: the first layer is freed correctly, the second layer currently is not. `Vec<Rect>` in HOG is fine since it only has one layer.

The following `cv_vec_of_points_drop` has the right fix.

```
void cv_vec_of_points_drop(VecPoints* pvs) {
    if (pvs->array != nullptr) {
        // Free internal VecPoints
        for (size_t i = 0; i < pvs->size; i++) {
            cv_vec_of_point_drop(&pvs->array[i]);
        }
        free(pvs->array);
        pvs->array = nullptr;
        pvs->size = 0;
    }
}
```

Reply:
@nebgnahz yep, I had this code but I deleted it for some reason. You convienced me.

However, there is a proble. You see, `std::vector` is never freed. We copy `std::Vector<std::Vector<Point2i>>` to `CVecOfPoints`, but we don't free the source. Your code is showing this problem: fixed `drop` won't save us from this memory leak.

Reply:
Yes it is a memory copy (so expensive); but `std::vector` is freed when it goes out of scope (that's C++). I ran the test code with the fix, the memory is kept around 40M~50M on OS X after running for at least 1 minutes.

Reply:
@nebgnahz done